File: travel-backend-java/src/main/java/com/travel/backend/common/SeasonPrice.java
package com.travel.backend.common;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.time.MonthDay;

/**
 * 酒店季节定价：与 travel-agent-python app/common/season.py 保持完全一致的规则。
 */
public final class SeasonPrice {

    public static final BigDecimal PEAK_FACTOR = new BigDecimal("1.8");   // 节假日窗口
    public static final BigDecimal SUMMER_FACTOR = new BigDecimal("1.5"); // 暑期 7-8 月
    public static final BigDecimal OFF_FACTOR = new BigDecimal("0.85");   // 淡季 12/1/2 月
    public static final BigDecimal NORMAL_FACTOR = BigDecimal.ONE;

    private SeasonPrice() {
    }

    public static BigDecimal factor(LocalDate d) {
        if (d == null) {
            d = LocalDate.now();
        }
        if (inHoliday(d)) {
            return PEAK_FACTOR;
        }
        int m = d.getMonthValue();
        if (m == 7 || m == 8) {
            return SUMMER_FACTOR;
        }
        if (m == 12 || m == 1 || m == 2) {
            return OFF_FACTOR;
        }
        return NORMAL_FACTOR;
    }

    public static String label(LocalDate d) {
        if (d == null) {
            d = LocalDate.now();
        }
        if (inHoliday(d)) {
            return "节假日旺季";
        }
        int m = d.getMonthValue();
        if (m == 7 || m == 8) {
            return "暑期旺季";
        }
        if (m == 12 || m == 1 || m == 2) {
            return "淡季";
        }
        return "平季";
    }

    public static BigDecimal apply(BigDecimal base, LocalDate d) {
        return base.multiply(factor(d)).setScale(2, RoundingMode.HALF_UP);
    }

    private static boolean inHoliday(LocalDate d) {
        MonthDay md = MonthDay.from(d);
        return between(md, MonthDay.of(1, 24), MonthDay.of(2, 8))
                || between(md, MonthDay.of(4, 30), MonthDay.of(5, 5))
                || between(md, MonthDay.of(9, 30), MonthDay.of(10, 7));
    }

    private static boolean between(MonthDay md, MonthDay from, MonthDay to) {
        return !md.isBefore(from) && !md.isAfter(to);
    }
}

File: travel-agent-python/app/services/season_price.py
"""酒店季节定价：必须与 `app/common/season.py`（Agent 侧）和 Java `SeasonPrice` 三方同规则。

规则是"日历窗口"而非真实节假日表：春节窗口 1/24-2/8、五一 4/30-5/5、国庆 9/30-10/7。
任何一侧改动都会让预算合计与 Agent 生成的报价对不上，故三个实现都要同步。
"""

from __future__ import annotations

from datetime import date, time
from decimal import ROUND_HALF_UP, Decimal

PEAK_FACTOR = Decimal("1.8")
SUMMER_FACTOR = Decimal("1.5")
OFF_FACTOR = Decimal("0.85")
NORMAL_FACTOR = Decimal("1")

_HOLIDAY_WINDOWS = ((time(1, 24), time(2, 8)), (time(4, 30), time(5, 5)), (time(9, 30), time(10, 7)))


def _month_day(value: date) -> time:
    # 用 time(month, day) 当 MonthDay 用：只比较月/日，且能正确处理跨年窗口外的比较
    return time(value.month, value.day)


def _in_holiday(value: date) -> bool:
    md = _month_day(value)
    return any(start <= md <= end for start, end in _HOLIDAY_WINDOWS)


def factor(start_date: date | None = None) -> Decimal:
    day = start_date or date.today()
    if _in_holiday(day):
        return PEAK_FACTOR
    if day.month in (7, 8):
        return SUMMER_FACTOR
    if day.month in (12, 1, 2):
        return OFF_FACTOR
    return NORMAL_FACTOR


def label(start_date: date | None = None) -> str:
    day = start_date or date.today()
    if _in_holiday(day):
        return "节假日旺季"
    if day.month in (7, 8):
        return "暑期旺季"
    if day.month in (12, 1, 2):
        return "淡季"
    return "平季"


def apply(base: Decimal, start_date: date | None = None) -> Decimal:
    return (Decimal(base) * factor(start_date)).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
